### ISSUE TITLE
__len__, __getitem__ and __data_generation gets called multiple times 

### ISSUE THREAD

**Issue by mu-mandayam** [keras-team/keras#15701]
I have a large, non-image, CSV file that I want to read in batches and feed to model.fit_generator. I have written a DataGenerator(keras.utils.all_utils.Sequence) class with the following methods:
__init__,
__len__,
__getitem__,
on_epoch_end, and
__data_generation

I instantiate a training and validation class and then call model.fit_generator with the correct batch size and other parameters. I noticed that __len__, __getitem__, and __data_generation gets called multiple times and out of sequence throwing off my modeling accuracy. For example, if my total rows in CSV is 350 and I set my batch size to 50, __getitem__ and __data_generation should be called 7 times each for training and validation data sets. Instead, I see that the order of reading the data is not sequential and __getitem__ and __data_generation are called more than 7 times. I have set shuffle=False.

I am using the following:
NVIDIA GeForce GTX 1660 Ti
keras 2.6.0
keras.utils 1.0.13
tensorflow 2.6.0
linux 5.11.0-38
python 3.9.5

Here's my sample code:

class DataGenerator(keras.utils.all_utils.Sequence):
#  
    def __init__(self, file_name, rows_per_batch=50, shuffle=False):
        self.rows_per_batch = rows_per_batch
        self.shuffle = shuffle
        self.file_name = file_name
        self.on_epoch_end()
        reader = csv.reader(open(self.file_name, 'r'))
        self.lines = list(reader)

    def __len__(self):
        return len(self.lines) // self.rows_per_batch

    def __getitem__(self, index):
        skip_rows = index * self.rows_per_batch
        nrows = (index + 1) * self.rows_per_batch
        X, y = self.__data_generation(self.folder_name, self.file_name, skip_rows, nrows)
        
        return X, y

    def on_epoch_end(self):
        if self.shuffle == True:
            np.random.shuffle(self.indexes)

    def __data_generation(self, file_name, skip_rows, nrows, config_file):
       
        df = pd.read_csv(file_name, header=None, skiprows=skip_rows, nrows=nrows)

        < do some data processing here >
        
        return X, y

# main program

    file_name = <some filename>
    rows_per_batch = <some number>
    
    a = DataGenerator(file_name, rows_per_batch, shuffle=False)
    b = DataGenerator(file_name, rows_per_batch, shuffle=False)
    
    # Train model on dataset
    model.fit_generator(generator=a,
                        validation_data=b,
                        use_multiprocessing=False,
                        shuffle=False,
                        epochs=10,
                        workers=6)      # workers is the number of threads. kind of a hyper parameter for datagenerator class
    
Please go to TF Forum for help and support:

https://discuss.tensorflow.org/tag/keras

If you open a GitHub issue, here is our policy:

It must be a bug, a feature request, or a significant problem with the documentation (for small docs fixes please send a PR instead).
The form below must be filled out.

**Here's why we have that policy:**.

Keras developers respond to issues. We want to focus on work that benefits the whole community, e.g., fixing bugs and adding features. Support only helps individuals. GitHub also notifies thousands of people when issues are filed. We want them to see you communicating an interesting problem, rather than being redirected to Stack Overflow.

**System information**.
- Have I written custom code (as opposed to using a stock example script provided in Keras):
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04):
- TensorFlow installed from (source or binary):
- TensorFlow version (use command below):
- Python version:
- Bazel version (if compiling from source):
- GPU model and memory:
- Exact command to reproduce:

You can collect some of this information using our environment capture script:

https://github.com/tensorflow/tensorflow/tree/master/tools/tf_env_collect.sh

You can obtain the TensorFlow version with:
python -c "import tensorflow as tf; print(tf.version.GIT_VERSION, tf.version.VERSION)"

**Describe the problem**.

Describe the problem clearly here. Be sure to convey here why it's a bug in Keras or why the requested feature is needed.

**Describe the current behavior**.

**Describe the expected behavior**.

**[Contributing](https://github.com/keras-team/keras/blob/master/CONTRIBUTING.md)**.

- Do you want to contribute a PR? (yes/no):
- If yes, please read [this page](https://github.com/keras-team/keras/blob/master/CONTRIBUTING.md) for instructions
- Briefly describe your candidate solution(if contributing):

**Standalone code to reproduce the issue**.

Provide a reproducible test case that is the bare minimum necessary to generate
the problem. If possible, please share a link to Colab/Jupyter/any notebook.

**Source code / logs**.

Include any logs or source code that would be helpful to diagnose the problem. If including tracebacks, please include the full traceback. Large logs and files should be attached. Try to provide a reproducible test case that is the bare minimum necessary to generate the problem.

**Comment by mu-mandayam** [comment]
Hi @sanatmpa1 - Thanks for taking up this issue. Is there anything wrong with the way I have implemented the class?

**Comment by chenmoneygithub** [comment]
Hi, could you provide the logging for this issue? Like how many times those methods are called, and where they are getting called. 

If possible, could you provide a reproducable example, so we can look closer into it, thanks!
